feat: Operation View Persistence

annapurna-gupta opened the pull request:
**Purpose of PR?**:

Extend Database (mslib/mscolab/models.py):
Extend the database (mslib/mscolab/models.py) to include a new ViewSettings table for persistent storage of view configurations.
Storing configurations in the database ensures they are retained and accessible across sessions and devices.

Implement server API Endpoints:
The API will act as a bridge between the client and the database to enable saving and restoring view configurations.

The key changes proposed include:

A way to save view configuration:
The client will send view configuration to the server via POST request.
The server will validate the data and save it to the viewSetting table in the database.

A way to restore configurations:
The client sends a GET request to fetch stored view settings.
The server retrieves configurations from ViewSettings and returns them to the client.

annapurna-gupta committed:
Add function to save view settings for an operation in database

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -36,6 +36,9 @@
 import os
 import re
 import sys
+import fs
```
we removed fs

ReimarBauer commented on mslib/mscolab/server.py:
```diff
@@ -271,8 +272,16 @@ def register_user(email, password, username, fullname):
 def verify_user(func):
     @functools.wraps(func)
     def wrapper(*args, **kwargs):
+        auth_header = request.headers.get('Authorization')
```
interesting addition, how do we use this?

ReimarBauer commented on mslib/mscolab/server.py:
```diff
@@ -872,6 +881,44 @@ def reset_request():
         return render_template('errors/403.html'), 403
 
+@APP.route("/save_operation_view_settings", methods=["POST"])
+def save_operation_view_settings():
```
allt the db handling should be done in the module file_manager

ReimarBauer commented on mslib/mscolab/server.py:
```diff
@@ -872,6 +881,44 @@ def reset_request():
+            logging.error("No data provided in request")
+            return jsonify({"success": False, "message": "No data provided"}), 400
+
+        op_id = settings_data.get("settings", {}).get("global", {}).get("op_id")
```
we maybe should verify with the active op_id and check that the user is a member of that.


When we use only settings_data how did we find out that the data was correct?

ReimarBauer commented on mslib/mscolab/server.py:
```diff
@@ -872,6 +881,44 @@ def reset_request():
         logging.warning("To send emails, the value of `MAIL_ENABLED` in `conf.py` should be set to True.")
         return render_template('errors/403.html'), 403
 
+@APP.route("/save_operation_view_settings", methods=["POST"])
```
this needs a `@verify_user` or anyone can push some data on the server

annapurna-gupta committed:
added logic for restoring

ReimarBauer commented on mslib/mscolab/server.py:
```diff
@@ -899,26 +890,51 @@ def save_operation_view_settings():
+@verify_user
+def get_operation_view_settings():
+    logging.info("in the server")
+    op_id = request.args.get('op_id', type=int)
```
we need a check 

if fm.is_member(user.id, op_id): 

so that not any user can request any view_settings, only those where he is member

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -559,7 +558,9 @@ def add_plugins(self):
 
     @QtCore.pyqtSlot(int)
     def activate_operation_slot(self, active_op_id):
+        self.send_view_settings_to_server()
```
maybe add a docstring something like "saves the recent settings, switches afterwards the operation and loads existing settings" ?

ReimarBauer commented on mslib/mscolab/server.py:
```diff
@@ -899,26 +890,51 @@ def save_operation_view_settings():
-            logging.debug("Updating existing ViewSettings record: u_id=%s, op_id=%s", u_id, op_id)
         else:
             view_setting = ViewSettings(op_id=op_id, u_id=u_id, settings=settings_str)
             db.session.add(view_setting)
```
all db code has to be a method in the file_manager module

the server itself should only do Post or Get routes

ReimarBauer reviewed (commented):
when I do a file quit it should quit without loading the view settings from the first Flight Track

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1301,6 +1314,186 @@ def update_flight_track_settings(self, flight_track, view=None, remove=False, up
                               getattr(view, 'name', 'unknown'), type(view).__name__,
                               getattr(view, 'view_id', 'unknown'), str(ex))
 
+    def get_operation_view_settings(self):
```
there is also a module msui.mscolab where we have the other communications of mscolab with the server.

There we use something like requests.Session(), see
https://github.com/Open-MSS/MSS/blob/develop/mslib/msui/mscolab.py#L460

or another example by using data
https://github.com/Open-MSS/MSS/blob/develop/mslib/msui/mscolab_admin_window.py#L202

matrss commented:
The failing migration tests come from the fact that your database model declares the ViewSettings.settings column as text that is not nullable, while your migration declares it as json which is nullable. So there is a discrepancy between the model and the database schema, which the test correctly identifies as an issue and therefore fails.

annapurna-gupta committed:
added test and logic for reapply setting for an operation

ReimarBauer commented on mslib/msui/mscolab.py:
```diff
@@ -1747,6 +1747,8 @@ def archive_operation(self, _):
     @verify_user_token
     def set_active_op_id(self, item):
+        from mslib.msui.msui_mainwindow import MSUIMainWindow
+        msui = MSUIMainWindow()
```
this is not intended to be done that way. because we have a msui window running

I think you need to move your code into this module

ReimarBauer reviewed (changes requested):
please move all new mscolab view_settings function into mslib.msui.mscolab

When you want to use them in the msui_mainwindow you can reach them by e.g.
self.mscolab.load_operation_view_settings


then you can also add the methods to the places where you currenbtly created a new instance 

keep in msui_mainwindow those which have not a relation to mscolab only or the qt interface

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1296,10 +1303,197 @@ def update_flight_track_settings(self, flight_track, view=None, remove=False, up
+                     settings.get("global", {}).get("op_id"), self.mscolab.mscolab_server_url)
+
+        try:
+            api_endpoint = f"{self.mscolab.mscolab_server_url}/save_operation_view_settings"
```
This needs to be changed to something like 
self.conn.request_post after you moved it to mslib.msui.mscolab

annapurna-gupta committed:
Move operation view settings functions to mscolab.py

ReimarBauer commented on mslib/mscolab/server.py:
```diff
@@ -873,6 +873,55 @@ def reset_request():
+            return jsonify({"success": False, "message": "No data provided"}), 400
+
+        op_id = settings_data.get("settings", {}).get("global", {}).get("op_id")
+        if not op_id:
```
is not None?

ReimarBauer replied:
dont' abuse boolean checks, see

https://github.com/Open-MSS/MSS/issues/1614

ReimarBauer commented on mslib/mscolab/server.py:
```diff
@@ -873,6 +873,55 @@ def reset_request():
+@verify_user
+def get_operation_view_settings():
+    logging.info("Fetching view settings for operation")
+    op_id = request.args.get('op_id', type=int)
```
you need request.form.get('op_id')

annapurna-gupta committed:
imporved communication

git add mslib/msui/msui_mainwindow.py
git rebase --continue
B
B
B
B
A
A
A
A
A
A
A
A
A
A
A
A
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
B
A
A

B

ReimarBauer commented on mslib/mscolab/file_manager.py:
```diff
@@ -778,3 +779,63 @@ def import_permissions(self, import_op_id, current_op_id, u_id):
+        """Save view settings for an operation and user to the database."""
+
+        if not self.is_member(user.id, op_id):
+            logging.error("User %s is not a member of operation %s", user.id, op_id)
```
and not is_viewer?

matrss commented on mslib/mscolab/file_manager.py:
```diff
@@ -778,3 +779,63 @@ def import_permissions(self, import_op_id, current_op_id, u_id):
+            return False, "Access denied: User is not a member of this operation", {}
+
+        try:
+            settings_exist = db.session.query(ViewSettings).filter_by(op_id=op_id).first() is not None
+            if not settings_exist:
+                return True, "No settings found for this operation", {"views": [], "global": {}}
```
What is the reason for this preliminary check, when you are just interested in the settings for (op_id, user)?

matrss commented on mslib/mscolab/file_manager.py:
```diff
@@ -778,3 +779,63 @@ def import_permissions(self, import_op_id, current_op_id, u_id):
+            if settings is None:
+                return True, "No view settings found", {"views": [], "global": {}}
+
+            if not isinstance(settings, dict):
```
Can settings ever be a dict after reading it from the database?

matrss commented on mslib/mscolab/models.py:
```diff
@@ -222,3 +222,24 @@ def __init__(self, op_id, u_id, commit_hash, version_name=None, comment=None):
+class ViewSettings(db.Model):
+
+    __tablename__ = "viewsettings"
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)  # noqa: A003
+    op_id = db.Column(db.Integer, db.ForeignKey('operations.id'))
+    u_id = db.Column(db.Integer, db.ForeignKey('users.id'))
```
IMO (op_id, u_id) should build a composite primary key, instead of having an additional third id. Then the unique constraint on (op_id, u_id) would be automatic.

matrss commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1027,7 +1040,7 @@ def create_view_handler(self, _type):
 
     def restore_views_for_active_flighttrack(self):
-        if not self.active_flight_track:
+        if not self.active_flight_track or self.mscolab.active_op_id:
```
```suggestion
        if not (self.active_flight_track or self.mscolab.active_op_id):
```

Otherwise this means "if there is no active flight track or if there is a active op id", which doesn't seem right.

matrss commented on tests/_test_mscolab/test_migrations.py:
```diff
@@ -96,6 +96,7 @@ def test_upgrade_from(revision, iterations, mscolab_app, tmp_path):
         metadata.reflect(bind=db.engine)
         expected_data = {name: db.session.execute(table.select()).all() for name, table in metadata.tables.items()}
         del expected_data["alembic_version"]  # the alembic_version table will be different, but that is expected
+        expected_data["viewsettings"] = []
```
There should be no need to mess with expected_data here, it is populated with the correct content in line 97.

matrss commented on tests/_test_mscolab/test_migrations.py:
```diff
@@ -114,6 +115,8 @@ def test_upgrade_from(revision, iterations, mscolab_app, tmp_path):
             flask_migrate.check(directory=migrations_path)
             actual_data = {name: db.session.execute(table.select()).all() for name, table in db.metadata.tables.items()}
+            if 'viewsettings' not in actual_data:
+                actual_data['viewsettings'] = []
```
This defeats the purpose of the test. Setting `expected_data["viewsettings"]` and `actual_data["viewsettings"]` to [] and then comparing them for equality is a no-op.

matrss commented on tests/_test_mscolab/test_migrations.py:
```diff
@@ -128,6 +131,9 @@ def test_upgrade_from(revision, iterations, mscolab_app, tmp_path):
             del actual_data_after_downgrade["alembic_version"]  # expected data doesn't have the revision table
+            if 'viewsettings' not in actual_data_after_downgrade:
+                actual_data_after_downgrade['viewsettings'] = []
+            expected_data["viewsettings"] = []
```
Same as before, this defeats the purpose of the test. `actual_data_after_downgrade` should, by definition, not contain viewsettings, because it should reflect the state of the schema back in MSS v8/v9.

annapurna-gupta committed:
Update mslib/msui/msui_mainwindow.py

Co-authored-by: Matthias Riße <9308656+matrss@users.noreply.github.com>

annapurna-gupta committed:
added test

annapurna-gupta committed:
added suggestion fix

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -597,6 +597,13 @@ def add_plugin_submenu(self, name, extension, function, pickertype, plugin_type=
 
     def update_treewidget_op_fl(self, op_fl, flight):
+
+        # if self.mscolab.active_op_id is not None:
```
why is that comment off?

ReimarBauer commented on mslib/msui/socket_control.py:
```diff
@@ -241,11 +241,12 @@ def disconnect(self):
         self.sio.disconnect()
 
-    def request_post(self, api, data=None, files=None):
+    def request_post(self, api, data=None, files=None, headers=None):
```
for what is headers needed here?

ReimarBauer reviewed (commented):
see comments,

please close resolved items

annapurna-gupta committed:
removed id from db cloumn'

annapurna-gupta committed:
Removed id column and updated migration

annapurna-gupta committed:
chnages in msui/mscolab.py

ReimarBauer commented on mslib/msui/mpl_qtwidget.py:
```diff
@@ -494,11 +494,11 @@ def redraw_yaxis(self):
         axes_label_size = (self.sideview_size_settings["axes_label_size"]
                            if self.settings["axes_label_size"] == "default"
-                           else int(self.settings["axes_label_size"]))
+                           else int(self.settings["axes_label_size"].replace("pt", "").strip()))
```
interesting, how did you found that. Maybe we have to fix that in stable too.

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1027,6 +1028,7 @@ def create_view_handler(self, _type):
 
     def restore_views_for_active_flighttrack(self):
+        if not (self.active_flight_track or self.mscolab.active_op_id):
         if not self.active_flight_track:
```
this looks like a merge issue, line 1032 is duplicated and not intended

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1266,9 +1269,14 @@ def status(self):
 
+    def update_flight_track_settings(self, flight_track, view=None, remove=False):
+        """Update the flight_track_settings dictionary when a flight track or view is created, modified, or removed."""
+        if self.mscolab.active_op_id:
```
this looks wrong.

What should happen after the "if"?

def ... duplicated?

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1307,6 +1315,95 @@ def update_flight_track_settings(self, flight_track, view=None, remove=False, up
+                else:
+                    logging.debug("No setting available for %s", json_key)
+            self.activated_flight_tracks.clear()
             except Exception as ex:
```
where is the try: ?

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1325,13 +1422,17 @@ def closeEvent(self, event):
                 self.mscolab.logout()
+
             # Table View stick around after MainWindow closes - maybe some dangling reference?
-            # This removes them for sure!
```
the comment maybe important

annapurna-gupta committed:
fixes lint test

ReimarBauer reviewed (changes requested):
I did 

pixi shell -e dev
pixi run mscolab db --seed
pixi run mscolab start

then logged in as b@notexisting.org
and activate operation one
afterwards I opened 5 views

<img width="739" height="775" alt="Bildschirmfoto 2025-09-01 um 13 06 07" src="https://github.com/user-attachments/assets/bdfe08c9-683e-440d-bec0-55b56be1d37b" />

then I switched to operation "two" and created a topview
<img width="739" height="780" alt="Bildschirmfoto 2025-09-01 um 13 10 34" src="https://github.com/user-attachments/assets/1a1d26ec-2b95-4bf4-89e3-d42ff4d8c0d9" />

Now I have 2 different setups stored
I toggled between both operations until I recognized that I lost two views.
All topviews are not shown
<img width="730" height="769" alt="Bildschirmfoto 2025-09-01 um 13 14 40" src="https://github.com/user-attachments/assets/84f0b6bd-bf01-4404-961f-c895f3a8ad81" />

Please have a look on this


first time cycling one - two it removes 1 top view. By 2 cycles all are removed

ReimarBauer commented:
on quit the UI shows 

```
Error message: The truth value of an empty array is ambiguous. Use `array.size > 0` to check that an array is not empty..
Hint: Check used waypoints..
ERROR: MapPrefetcher Exception <class 'owslib.util.ServiceException'> - The data corresponding to your request is not available. Please check the times and/or path you have specified.
```
it should just quit

ReimarBauer commented:
Also on quit it seems to have closed only the mscolab connection but don't have closed the flight tracks in the list.

Because a flight track is then activated it tries to map the open views and saves them. 
We should not save view_settings for not stored flight tracks. 
I guess we see this in the waypoint table. 
And on quit we should close open flight tracks and when we come from mscolab nothing should get into a flight track by a quit.

ReimarBauer commented:
On a logout (profile picture, logout) it stores the kind of view but not its parameters. e.g. if I select a topview with data from open-mss.org 700 hPa or 300 hPa. After relogin and selecting the operation it did not use a server url. It just openes in my case 2 x topview.

annapurna-gupta commented:
> I did
> 
> pixi shell -e dev pixi run mscolab db --seed pixi run mscolab start
> 
> then logged in as [b@notexisting.org](mailto:b@notexisting.org) and activate operation one afterwards I opened 5 views
> 
> <img alt="Bildschirmfoto 2025-09-01 um 13 06 07" width="739" height="775" src="https://private-user-images.githubusercontent.com/278728/484180616-bdfe08c9-683e-440d-bec0-55b56be1d37b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY3Mzc1NTksIm5iZiI6MTc1NjczNzI1OSwicGF0aCI6Ii8yNzg3MjgvNDg0MTgwNjE2LWJkZmUwOGM5LTY4M2UtNDQwZC1iZWMwLTU1YjU2YmUxZDM3Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMVQxNDM0MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MGM1MTViZmVhZjMyMWY2YjAzOThhYzk1MDJiNDkzMzRjNjNlYWUwZjZlZjBhY2QyNTAwNGFjODIyMDE5YThhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.qg_r5pwlnYQ7EaHciEnNJKLxj_Xma7-XSiMnPoWcVsc">
> then I switched to operation "two" and created a topview <img alt="Bildschirmfoto 2025-09-01 um 13 10 34" width="739" height="780" src="https://private-user-images.githubusercontent.com/278728/484180952-1a1d26ec-2b95-4bf4-89e3-d42ff4d8c0d9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY3Mzc1NTksIm5iZiI6MTc1NjczNzI1OSwicGF0aCI6Ii8yNzg3MjgvNDg0MTgwOTUyLTFhMWQyNmVjLTJiOTUtNGJmNC04OWUzLWQ0MmZmNGQ4YzBkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMVQxNDM0MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NjljMzg1Y2FiMTkyNTU1YWQ2ZTIxYTEyMjI3NTIwN2Y3ZDE0ZjA4ODU3NTFhZTFlNjg2NmE1MWEyMWI0Nzk3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.sa8KzcIaB5CdvAPk92cvCgKIGHbZw_A-dUOvNF_ueME">
> 
> Now I have 2 different setups stored I toggled between both operations until I recognized that I lost two views. All topviews are not shown <img alt="Bildschirmfoto 2025-09-01 um 13 14 40" width="730" height="769" src="https://private-user-images.githubusercontent.com/278728/484182180-84f0b6bd-bf01-4404-961f-c895f3a8ad81.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY3Mzc1NTksIm5iZiI6MTc1NjczNzI1OSwicGF0aCI6Ii8yNzg3MjgvNDg0MTgyMTgwLTg0ZjBiNmJkLWJmMDEtNDQwNC05NjFmLWM4OTVmM2E4YWQ4MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMVQxNDM0MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MTliZDVhNWQ2ZmE5OTc2MzhlMWJkYWMwYTJhMWZjNThmYTA2ZjIzYjE1OTEyZmQ3ZjYzODkxYWViZmY5MjVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2V-Zsz-yRTn0z4DuGda60bkZlIjmwpzJgNlwag_rtD0">
> 
> Please have a look on this
> 
> first time cycling one - two it removes 1 top view. By 2 cycles all are removed

so at this time...configuration were enabled?

annapurna-gupta commented:
> I did
> 
> pixi shell -e dev pixi run mscolab db --seed pixi run mscolab start
> 
> then logged in as [b@notexisting.org](mailto:b@notexisting.org) and activate operation one afterwards I opened 5 views
> 
> <img alt="Bildschirmfoto 2025-09-01 um 13 06 07" width="739" height="775" src="https://private-user-images.githubusercontent.com/278728/484180616-bdfe08c9-683e-440d-bec0-55b56be1d37b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY3MzkzMjYsIm5iZiI6MTc1NjczOTAyNiwicGF0aCI6Ii8yNzg3MjgvNDg0MTgwNjE2LWJkZmUwOGM5LTY4M2UtNDQwZC1iZWMwLTU1YjU2YmUxZDM3Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMVQxNTAzNDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNmIyZjIyYjlhZGI0OWQ3MThlNzg2MDNhZjhiYWYzNTkzM2VmYjZmNWE1Njk5YzQ5YmJiYTk0NWJiYTYzYTk5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Fjcv1YJo8o5LiRs-EdCX58F1FYwjxAwKagzscj1dZT4">
> then I switched to operation "two" and created a topview <img alt="Bildschirmfoto 2025-09-01 um 13 10 34" width="739" height="780" src="https://private-user-images.githubusercontent.com/278728/484180952-1a1d26ec-2b95-4bf4-89e3-d42ff4d8c0d9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY3MzkzMjYsIm5iZiI6MTc1NjczOTAyNiwicGF0aCI6Ii8yNzg3MjgvNDg0MTgwOTUyLTFhMWQyNmVjLTJiOTUtNGJmNC04OWUzLWQ0MmZmNGQ4YzBkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMVQxNTAzNDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00N2Q5NTg1NDA1ZDU2ZGY4YTc0MjU1MDkyOGRjNWMwMDAzYzFiZGUyYjVkM2I5Nzg5OTk2ZWZlYmM3MDJkOWUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.6fMuCwMqAa1A9ctCaTenF9mpVank7juWSVHllO9ooaU">
> 
> Now I have 2 different setups stored I toggled between both operations until I recognized that I lost two views. All topviews are not shown <img alt="Bildschirmfoto 2025-09-01 um 13 14 40" width="730" height="769" src="https://private-user-images.githubusercontent.com/278728/484182180-84f0b6bd-bf01-4404-961f-c895f3a8ad81.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY3MzkzMjYsIm5iZiI6MTc1NjczOTAyNiwicGF0aCI6Ii8yNzg3MjgvNDg0MTgyMTgwLTg0ZjBiNmJkLWJmMDEtNDQwNC05NjFmLWM4OTVmM2E4YWQ4MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMVQxNTAzNDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMGIwMTEwY2E1MGI4YjgyZmNmYzAyNGE4YjY5NjgyMDU2MzE1ZGJjNWE2NjQ3MDdhZjhkMDljZTNjNzQxZTVkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.CRwBYs_xkNZ_JerNMja-VGqKtqEdjYe_DtIHJPJvuAQ">
> 
> Please have a look on this
> 
> first time cycling one - two it removes 1 top view. By 2 cycles all are removed

[Screencast from 2025-09-01 20-26-36.webm](https://github.com/user-attachments/assets/b573d9ee-b973-43b6-b178-a57899ebd104)
see here i have created 
1. "aa" with 5 top view
2. "bb"  with one top view
and even after multiple switching the views are restored correctly

annapurna-gupta committed:
enhance logic

ReimarBauer commented:
> > I did
> > pixi shell -e dev pixi run mscolab db --seed pixi run mscolab start
> > then logged in as [b@notexisting.org](mailto:b@notexisting.org) and activate operation one afterwards I opened 5 views
> > <img alt="Bildschirmfoto 2025-09-01 um 13 06 07" width="739" height="775" src="https://private-user-images.githubusercontent.com/278728/484180616-bdfe08c9-683e-440d-bec0-55b56be1d37b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY3Mzc1NTksIm5iZiI6MTc1NjczNzI1OSwicGF0aCI6Ii8yNzg3MjgvNDg0MTgwNjE2LWJkZmUwOGM5LTY4M2UtNDQwZC1iZWMwLTU1YjU2YmUxZDM3Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMVQxNDM0MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MGM1MTViZmVhZjMyMWY2YjAzOThhYzk1MDJiNDkzMzRjNjNlYWUwZjZlZjBhY2QyNTAwNGFjODIyMDE5YThhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.qg_r5pwlnYQ7EaHciEnNJKLxj_Xma7-XSiMnPoWcVsc">
> > then I switched to operation "two" and created a topview <img alt="Bildschirmfoto 2025-09-01 um 13 10 34" width="739" height="780" src="https://private-user-images.githubusercontent.com/278728/484180952-1a1d26ec-2b95-4bf4-89e3-d42ff4d8c0d9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY3Mzc1NTksIm5iZiI6MTc1NjczNzI1OSwicGF0aCI6Ii8yNzg3MjgvNDg0MTgwOTUyLTFhMWQyNmVjLTJiOTUtNGJmNC04OWUzLWQ0MmZmNGQ4YzBkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMVQxNDM0MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NjljMzg1Y2FiMTkyNTU1YWQ2ZTIxYTEyMjI3NTIwN2Y3ZDE0ZjA4ODU3NTFhZTFlNjg2NmE1MWEyMWI0Nzk3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.sa8KzcIaB5CdvAPk92cvCgKIGHbZw_A-dUOvNF_ueME">
> > Now I have 2 different setups stored I toggled between both operations until I recognized that I lost two views. All topviews are not shown <img alt="Bildschirmfoto 2025-09-01 um 13 14 40" width="730" height="769" src="https://private-user-images.githubusercontent.com/278728/484182180-84f0b6bd-bf01-4404-961f-c895f3a8ad81.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY3Mzc1NTksIm5iZiI6MTc1NjczNzI1OSwicGF0aCI6Ii8yNzg3MjgvNDg0MTgyMTgwLTg0ZjBiNmJkLWJmMDEtNDQwNC05NjFmLWM4OTVmM2E4YWQ4MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkwMVQxNDM0MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MTliZDVhNWQ2ZmE5OTc2MzhlMWJkYWMwYTJhMWZjNThmYTA2ZjIzYjE1OTEyZmQ3ZjYzODkxYWViZmY5MjVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2V-Zsz-yRTn0z4DuGda60bkZlIjmwpzJgNlwag_rtD0">
> > Please have a look on this
> > first time cycling one - two it removes 1 top view. By 2 cycles all are removed
> 
> so at this time...configuration were enabled?

yes

annapurna-gupta committed:
added is_flight_track_stored function is msui_mainwindow.py

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1271,10 +1267,17 @@ def status(self):
+    def is_flight_track_stored(self, flight_track):
+        """Check if the active flight track is saved (has a valid filename)."""
+        if flight_track is None:
+            return
```
return with no argument returns None, do you want a boolean here?

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1271,10 +1267,17 @@ def status(self):
             return (f"Status : User Configuration '{constants.MSUI_SETTINGS}' loaded")
 
-    def update_flight_track_settings(self, flight_track, view=None, remove=False, update_global=False):
+    def is_flight_track_stored(self, flight_track):
```
this is not needing self? so maybe it can be moved to mslib.utils.view_restoration?

ReimarBauer commented on mslib/msui/mscolab.py:
```diff
@@ -1750,6 +1820,8 @@ def set_active_op_id(self, item):
         logging.debug('set_active_op_id %s %s %s', item, item.op_id, self.active_op_id)
         if not self.ui.local_active and item.op_id == self.active_op_id:
             return
+        # self.ui.save_view_settings()
```
is that needed?

annapurna-gupta committed:
added suggested chnages

ReimarBauer commented on mslib/msui/msui_mainwindow.py:
```diff
@@ -1267,15 +1271,15 @@ def status(self):
-        if flight_track is None:
-            return
-        return flight_track.filename is not None
+    # def is_flight_track_stored(self, flight_track):
```
when this is not needed here, please remove it

ReimarBauer reviewed (commented):
Do you want to work on this in the next PR?
changing to another operation did not save the used views.

https://github.com/user-attachments/assets/3c630aca-2b98-4f60-8ca0-08278d147f6c

ReimarBauer commented:
This maybe should be fixed in this PR, or you do work with the one above parallel in a further PR. (after we merged this one)

https://github.com/user-attachments/assets/e22d0ac3-de62-4cd6-a03f-f28c7758b6cc

ReimarBauer commented:
It seems changing an operatation views does not become saved on logout. This is likly because of the crash in 
```
File "/Users/reimarbauer/PycharmProjects/2025/annapurna-gupta/MSS/mslib/msui/tableview.py", line 412, in set_settings
    row, self.waypoints_model.column_index("flightlevel")
````
Also closing msui does not update

Do you want to work on this in parallel to the others after we merge this one?

https://github.com/user-attachments/assets/6cefbf19-9eb2-4c0d-9897-9104bc91505f

annapurna-gupta committed:
matching the header text

annapurna-gupta committed:
fixex table view error and code cleanup

annapurna-gupta commented:
> This maybe should be fixed in this PR, or you do work with the one above parallel in a further PR. (after we merged this one)
> 
>  showing_an_error_when_table_view_is_stored.mp4

likely it's done

annapurna-gupta commented:
> Do you want to work on this in the next PR? changing to another operation did not save the used views.
> 
>  not_saving_on_operation_change.mp4

it's done

annapurna-gupta committed:
save operation views on logout

annapurna-gupta committed:
store and restore views when switch btw operation (if restore_view is enable)

ReimarBauer commented:
when toggling between flight tracks and operations, it does not save for fligh tracks. flight tracks are saved when msui was quitted.

https://github.com/user-attachments/assets/20496dc1-8267-475b-a49e-11167c10a2ec

annapurna-gupta committed:
save flight track settings when switching to operation

ReimarBauer commented:
When starting with a flight track, removing of a view gets not stored when changing to an operation.

https://github.com/user-attachments/assets/09e11977-37ea-4e8e-ad6f-2164d4ab3c63

adding an other view works

annapurna-gupta commented:
> When starting with a flight track, removing of a view gets not stored when changing to an operation.
> 
>  flighttrack_to_operation_did_not_update.mp4 
> adding an other view works

yeah, i know about this one.. Currently, there’s no logic in the codebase to delete settings from the JSON file, so they don’t get removed. But if you freshly load a flight track, create views, remove views, and then switch to an operation, it works correctly.

ReimarBauer commented:
> > When starting with a flight track, removing of a view gets not stored when changing to an operation.
> > flighttrack_to_operation_did_not_update.mp4
> > adding an other view works
> 
> yeah, i know about this one.. Currently, there’s no logic in the codebase to delete settings from the JSON file, so they don’t get removed. But if you freshly load a flight track, create views, remove views, and then switch to an operation, it works correctly.

ok we merge the PR and you can improve it by using tests afterwards

ReimarBauer reviewed (approved):
There are still some questions which will be worked on in the next PR
